06-classification_continued.Rmd

```diff
@@ -559,16 +657,378 @@ The overall workflow for performing K-nearest neighbour classification using `ti
+        cancer_recipe <- recipe(as.formula(model_string), data = cancer_subset) %>%
+                          step_scale(all_predictors()) %>%
+                          step_center(all_predictors())
+
+        # tune the K-NN classifier with these predictors, and collect the accuracy for the best K
+        acc <- workflow() %>%
+          add_recipe(cancer_recipe) %>%
+          add_model(knn_spec) %>%
+          tune_grid(resamples = cancer_vfold, grid = 10) %>%
+          collect_metrics() %>%
+          filter(.metric == "accuracy") %>%
+          summarize(mx = max(mean))
+        acc <- acc$mx %>% unlist()
+
+        # add this result to the dataframe
+        accs[[j]] <- acc
+        models[[j]] <- model_string
+    }
+    jstar <- which.max(unlist(accs))
+    accuracies <- accuracies %>% add_row(size = i, model_string = models[[jstar]], accuracy = accs[[jstar]])
+    selected <- c(selected, names[[jstar]])
+    names <- names[-jstar]
+}
+accuracies
+```
+
+Interesting! The forward selection procedure first added the three meaningful variables `Perimeter`,
+`Concavity`, and `Smoothness`, followed by the irrelevant variables. Figure \@ref(fig:06-fwdsel-3)
+visualizes the accuracy as a function of the number of predictors in the model. You can see that
+as meaningful predictors are added, the estimated accuracy increases substantially; and as you add irrelevant
+variables, the accuracy either exhibits small fluctuations or decreases as the model attempts to tune the number
+of neighbours to account for the extra noise. In order to pick the right model from the sequence, you have 
+to trade off the desiderata to have a high accuracy, but also a simple model with few predictors (which has 
+a lower chance of overfitting). The way to perform that trade-off is to look for the *elbow*
+in Figure \@ref(fig:06-fwdsel-3), i.e., the place on the plot where the accuracy stops increasing dramatically and
+levels off or begins to decrease. The elbow in Figure \@ref(fig:06-fwdsel-3) appears to occur at the model with 
+3 predictors; after that point the accuracy levels off. So here the right trade-off of accuracy and number of predictors
+occurs with 3 variables: `Class ~ Perimeter + Concavity + Smoothness`. In other words, we have successfully removed irrelevant
+predictors from the model! It is always worth remembering, however, that what cross validation gives you 
+is an *estimate* of the true accuracy; you have to use your judgement when looking at this plot to decide
+where the elbow occurs, and whether adding a variable provides a meaningful increase in accuracy.
+
+> Remember: since the choice of which variables to include as predictors is
+> part of tuning your classifier, you *cannot use your test data* for this
+> process! 
+
+```{r 06-fwdsel-3, warning = FALSE, fig.retina = 2, out.width = "600", fig.cap = "Estimated accuracy versus the number of predictors for the sequence of models built using forward selection"}
```

Comment:
I don't think we need to show this code chunk to the learners.

Reply:
done